tools/haskell_module_worker/src/Compile.hs

```diff
@@ -0,0 +1,279 @@
+
+compile :: [String] -> Int -> CompileM Status
+compile flags workerVerbosity = do
+
+    when (workerVerbosity > 0) $
+      liftIO $ hPutStrLn stderr $
+        "Compiling with flags: " ++ unwords (map show flags)
+
+    let (cflags, flags') = partition ("-c" ==) flags
+    if null cflags then
+      return NonOneShotCompilation
+    else do
+      st <- compileOneShot flags' workerVerbosity
+      when (workerVerbosity > 0) $
+        liftIO $ hPutStrLn stderr $ "Finished with status: " ++ show st
+      return st
+
+-- | Compiles with the given flags and source files
+--
+-- Some of this function is copied from @ghc/Main.hs@.
+-- https://gitlab.haskell.org/ghc/ghc/-/blob/ghc-8.10.7-release/ghc/Main.hs#L151
+compileOneShot :: [String] -> Int -> CompileM Status
+compileOneShot flags workerVerbosity = CompileM $ ReaderT $ \dflags0 -> do
+
+    -- Parse flags
+    (dflags2, fileish_args, dynamicFlagWarnings) <-
+      parseDynamicFlags dflags0 (map noLoc flags)
+
+    -- Normalize paths
+    let
+      normalise_hyp fp
+        | strt_dot_sl && "-" `isPrefixOf` nfp = cur_dir ++ nfp
+        | otherwise                           = nfp
+        where
+# if defined(mingw32_HOST_OS)
+          strt_dot_sl = "./" `isPrefixOf` fp || ".\\" `isPrefixOf` fp
+# else
+          strt_dot_sl = "./" `isPrefixOf` fp
+# endif
+          cur_dir = '.' : [pathSeparator]
+          nfp = normalise fp
+      normal_fileish_paths = map (normalise_hyp . unLoc) fileish_args
+      (srcs, objs)         = partition_args normal_fileish_paths [] []
+
+    logsRef <- liftIO $ newIORef []
+
+    let
+      dflags3 = dflags2
+        { ldInputs = map (FileOption "") objs ++ ldInputs dflags2
+        , ghcMode = OneShot
+        , log_action = renderLog $ \msg -> do
+            when (workerVerbosity > 0) $
+              hPutStrLn stderr msg
+            modifyIORef logsRef (msg:)
+        }
+
+    _ <- setSessionDynFlags dflags3
+
+    st0 <- collectStatus logsRef $ handleFlagWarnings dflags3 dynamicFlagWarnings
+    if isErrorStatus st0 then
+      return st0
+    else
+      doCompile logsRef srcs
+
+-- Compiles the given source files after flags have been used
+-- to setup the session
+doCompile :: IORef [String] -> [(String, Maybe Phase)] -> Ghc Status
+doCompile logsRef srcs = do
+    let (hs_srcs, non_hs_srcs) = partition isHaskellishTarget srcs
+    if null (non_hs_srcs) then do
+      hsc_env <- GHC.getSession
+      collectStatus logsRef $
+        mapM_ (compileFile hsc_env StopLn) hs_srcs
+    else
+      return $ NonHaskellInputs $ map fst non_hs_srcs
+
+-- | Creates a status from the exceptions thrown by the given action
+-- and the logs in the given 'IORef'.
+collectStatus :: IORef [String] -> IO () -> Ghc Status
+collectStatus logsRef action = do
+    hsc_env <- GHC.getSession
+    liftIO $ do
+      ee <- try action
+      logs <- reverse <$> readIORef logsRef
+      case ee of
+        Left se -> case fromException se of
```

Comment:
I think the `fromException` calls here would be clearer with a type application/annotation.

Reply:
Done.